### ISSUE TITLE
UI::HtmlDialog - Big issue with show_modal on Mac

### ISSUE THREAD

**Issue by Fredosixx** [SketchUp/api-issue-tracker#524]
**Platform**: Mac OSX only
**SU versions**: All SU2017 to SU2020

Another big one...

Issue: `show_modal` blocks all code executed in a timer until the dialog is closed.

This can be tested with the following code, where 2 events (onload and onclick) display a message
  - via a synchronous `puts`
  - via a deferred `puts` via `UI.start_timer`

```ruby
def test_modal
	#HTML content
	html_text = ''
	html_text += %q(<!DOCTYPE HTML>)
	html_text += %q(<META http-equiv='Content-Script-Type' content='text/javascript'>)
	html_text += %q(<META http-equiv='Content-Type' Content='text/html; charset=UTF-8'>)
	html_text += '<SCRIPT>'
	html_text += "window.addEventListener('load', function() { sketchup.onload() ; }) ;"
	html_text += '</SCRIPT>'
	html_text += "<button onclick='sketchup.onclick()'>Click me</button>"
	
	#Creating the HtmlDialog
	hsh = { :dialog_title => 'Test modal', :preferences_key => 'toto_modal' }
	wdlg = UI::HtmlDialog.new hsh
	wdlg.add_action_callback("onload") do |action_context, p1, p2| 
		puts "ON onload interactive"
		UI.start_timer(0.5) { puts "ON LOAD in a timer" }
	end	
	wdlg.add_action_callback("onclick") do |action_context, p1, p2| 
		puts "ON CLICK interactive"
		UI.start_timer(0.5) { puts "ON CLICK in a timer" }
	end	
	wdlg.set_html html_text
	wdlg.show_modal
end
```

1) Launch `test_modal` from the Ruby console

2) The synchronous ONLOAD message appears, but the one executed in a timer

3) Same if you click on the button

4) Close the dialog. **All deferred messages are flushed**.

Considering that timers are quite used in dialog management and related code, this simply makes **`show_modal` unusable on Mac**.

[HtmlDialogTestModal.zip](https://github.com/SketchUp/api-issue-tracker/files/5146103/HtmlDialogTestModal.zip)

**Comment by DanRathbun** [comment]
**How does it perform (on Mac) when the dialog is normal (non-modal) ?**

(Works fine on Windows. This would be a "platform parity" issue.)

----

Saved into my test folder as:
```ruby
# Timer_within_modal_HtmlDialog_callback.rb
#
# fails on MacOS SketchUp

def test_modal

  javascript = %q[
      window.addEventListener('load', function() {
              sketchup.onload();
          }
      );
  ]

  # HTML content
  html_text = %[
  <!DOCTYPE html>
  <html>
      <head>
          <meta http-equiv='Content-Script-Type' content='text/javascript'>
          <meta http-equiv='Content-Type' Content='text/html; charset=UTF-8'>
          <script>
              #{javascript}
          </script>
      </head>
      <body>
          <button onclick='sketchup.onclick()'>Click me</button>
      </body>
  </html>
  ]

  # Creating the HtmlDialog
  wdlg = UI::HtmlDialog.new( dialog_title: 'Test modal' )

  wdlg.add_action_callback("onload") do |action_context, p1, p2| 
    puts "ON LOAD interactive"
    UI.start_timer(0.5) { puts "ON LOAD in a timer" }
  end	

  wdlg.add_action_callback("onclick") do |action_context, p1, p2| 
    puts "ON CLICK interactive"
    UI.start_timer(0.5) { puts "ON CLICK in a timer" }
  end	

  wdlg.set_html( html_text )

  wdlg.show_modal
end
```

**Comment by Fredosixx** [comment]
In non-modal mode, it works as expected.

It took me a while to realize the issue I had was related to the timers.

**Comment by DanRathbun** [comment]
Well, good catch.

**Comment by thomthom** [comment]
I suspect that would be the case of any modal dialog, not just `UI::HtmlDialog`.

**Comment by Fredosixx** [comment]
First, that's not the case on Windows.

On Mac with `UI::WebDialog`, the timer event is fired. But I understand that `UI::WebDialog` is not truly modal on Mac (the dialog is just put in front of others).

On Mac, the `UI.messagebox`, which is modal,  blocks the timer execution. That's not the case on Windows however. You can try with the following code
```ruby
def modal_messagebox
    msg_proc = proc { UI.start_timer(1) { puts "Timer message" } ; "interactive Message" }
    UI.messagebox msg_proc.call
end	
```

For me this is a bug. **Even for modal dialogbox**, and because everything is asynchronous in HTML-based dialogs, you need to have timers working when you have some Javascript driving the dialog and communication between Ruby and Javascript. For instance, just the proper sizing of the dialog box requires a timer on Mac.

**Comment by thomthom** [comment]
I suspect it's related to the platform. So I cannot tell if this immediately is a system limitation or not.

Requiring a timer for dialog communication sounds like it's related to workarounds for other defects limitations you have reported, correct? You wouldn't be needing this timer if the rest was addressed? (First time I've heard of timers being used for dialog communication.)

**Comment by Fredosixx** [comment]
I don't think so. This is independent of (numerous) bugs already logged.

On Mac, like on Windows, you need a timer to perform the calculation of dimensions and then the call to the Ruby `#resize` method. This is because at the `onload` event, you do not have correct sizes yet in Javascript. There are other needs for timer for many other purposes.

A modal dialog box is nothing else than a normal dialog box in terms of HTML and Javascript configuration. It can be as complex.

I understand this maybe a limitation on Mac (but remember it works fine on Windows). But then, it would mean that modal dialog box are unusable on Mac, except in very simple cases. At least this is worth that the HtmlDialog developer(s) on your side does some exploration and advise.

**Comment by thomthom** [comment]
This is a slight tangent, but I'm getting the impression that you are relying on pixel accurate adjustments for your webdialog content? Hence the need to pass dialog dimensions back and forth between Ruby and JS?

Have you tried making the content more flexible? It's common thing in web design since a site or web app can be viewed on such a wide range of monitors.

Do you have some screenshots of the type of webdialog content and layout you are working with?

**Comment by Fredosixx** [comment]
The main problem is the **vertical** dimensioning (horizontal is usual OK and can be done with HTML / CSS)
  - for simple dialog boxes, I adjust the size to a few pixels down the last element in the page
  - when you have variable-height elements, typical list or tables, I use the given dialog height, and instead adjust of the height of the list or table.

This is for initial sizing at load, and for manual sizing.

If you know other ways to that with a pure Html-CSS approach, thanks to advise.

**Comment by Fredosixx** [comment]
In my opinion, there are 2 different effects of a modal dialog box

1) It blocks the GUI of Sketchup (beep if you click outside the dialog)

2) It blocks the Ruby code execution at the `wdlg.show_modal`. 

This second effect is unnecessary, because `show_modal` return `nil` anyway (unlike `UI.inputbox` or `UI.messagebox`), and therefore you need a notification proc for transmitting the results back to Ruby.

We could maybe do **without** this second effect (like it is the case today with `UI::WebDialog` on Mac).

**Comment by thomthom** [comment]
> If you know other ways to that with a pure Html-CSS approach, thanks to advise.

We've been wanting to spend some time on making better UI/UX resources for extension developers.

If you could provide me some screenshots of existing dialogs you have we can use that as a use-case for making some examples. (Some of the WebDialogs where are facing problems porting to HtmlDialog)

While I do want to make improvements to HtmlDialog, I think we can provide examples that might unblock you with the currently released versions of SU+HtmlDialog. Something that can unblock you sooner rather than later.

**Comment by DanRathbun** [comment]
> The main problem is the vertical dimensioning (horizontal is usual OK and can be done with HTML / CSS)
> 
> * for simple dialog boxes, I adjust the size to a few pixels down the last element in the page
> * when you have variable-height elements, typical list or tables, I use the given dialog height, and instead adjust of the height of the list or table.
> This is for initial sizing at load, and for manual sizing.
> 
> If you know other ways to that with a pure Html-CSS approach, thanks to advise.

**Yes this is easily done with CSS.** You don't need to send sizes back to Ruby and have Ruby change the size.

I have an alpha Shortcut Lister I can show you both. It has some things to fix, and a few more features to add, but it is a good example of a list of data that can be changed and reloaded, and lets CSS do all the resizing as the user changes the width and/or height of the dialog window.

I'll start a private thread over at the SketchUp forums to discuss it (as an example) and post the alpha rbz.

**Comment by sketchupbot** [comment]
Logged as: SU-47092
